SplitFiles/customer.c

```diff
@@ -0,0 +1,168 @@
+        BreadState* bread = (BreadState*)params->message;
+        thisCustomer = base->rootCustomer->nextCustomer;
+        lastCustomer = base->rootCustomer;
+        while (thisCustomer != NULL) {
+            if (bread->type->letter == thisCustomer->breadOrder->letter) {
+                //We have a match
+                params->score += bread->type->cost * 2 //SHOCKING markup
+                         + 100 - abs(bread->toastedness - 100) //100 toastedness gets 100 bonus points
+                         + thisCustomer->ticksLeft;  //Speedy toast = points!
+                (params->slices) += 1;
+                params->messageAddress = 0;
+                params->message = NULL;
+                params->messageSourceAddress = NULL;
+                //Give back memory
+                free(bread);
+                //Remove customer
+                base->customerCount -= 1;
+                lastCustomer->nextCustomer = thisCustomer->nextCustomer;
+                free(thisCustomer);
+                redraw = 1;
+                //end loop
+                thisCustomer = NULL;
+            } else {
+                //Step on
+                lastCustomer = thisCustomer;
+                thisCustomer = thisCustomer->nextCustomer;
+            }
+        }
+        if (!redraw) {
+            // No customer for this toast - bad
+            params->score -= (bread->type->cost + bread->toastedness);  //+ wasted energy cost
+            params->messageAddress = 0;
+            params->message = NULL;
+            params->messageSourceAddress = NULL;
+            free(bread);
+        }
+
+    }
+
+    thisCustomer = base->rootCustomer->nextCustomer;
+    lastCustomer = base->rootCustomer;
+    while (thisCustomer != NULL) {
+        thisCustomer->ticksLeft -= 1;
+        if (thisCustomer->ticksLeft <= 0) {
+            //Customer leaves in disgust - reset links
+            params->score -= thisCustomer->breadOrder->cost;
+            base->customerCount -= 1;
+            lastCustomer->nextCustomer = thisCustomer->nextCustomer;
+            free(thisCustomer);
+            thisCustomer = lastCustomer;
+            redraw = 1;
+        }
+        lastCustomer = thisCustomer;
+        thisCustomer = thisCustomer->nextCustomer;
+    }
+
+    //Do we have a new customer? - this is independent of whether we have space
+    if (rand() % 150 == 0) {
+        //Yes we do!
+        if (base->customerCount < MAX_CUSTOMER_COUNT) {
+            //Add in a new one then!
+            newCustomer = malloc(sizeof(Customer));
+            newCustomer->breadOrder = rand_bread_type(params->breadBin);
+            newCustomer->nextCustomer = NULL;
+            newCustomer->ticksLeft = newCustomer->breadOrder->cost * (8 + rand() % 4);
+            redraw = 1;
+            base->customerCount += 1;
+            lastCustomer->nextCustomer = newCustomer;
+        } else {
+            //Something about reputation here - enhanced by lots of people waiting?
```

Comment:
could have a markup factor - dynamic pricing? Up for more waiting but down if things slow?

Reply:
I have this thought in the back of my mind about "reputation", which is different from score and would be a hidden variable (maybe revealed at the end of each day).

So things like good toast and quick service would increase reputation.  Bad toast and particularly unfilled orders would reduce it.  Someone turning up and finding the shop full I think should increase reputation ("ooohh that place is so popular, it must be good").

Whole point of it is that the frequency of new customers should be high for a good rep and low for a bad one.  So if you're really crap the game doesn't end it just gets very boring because no customers show up.

Thoughts?